lib/puppet/provider/package/yum.rb

```diff
@@ -142,13 +142,24 @@ def self.update_command
     'update'
   end
 
+  def makecache_fast
+    command = [command(:cmd), :makecache, :fast]
+    execute(command, :failonfail => false, :combine => false)
+  end
+
   def install
     wanted = @resource[:name]
     error_level = self.class.error_level
     update_command = self.class.update_command
     # If not allowing virtual packages, do a query to ensure a real package exists
     unless @resource.allow_virtual?
-      execute([command(:cmd), '-d', '0', '-e', error_level, '-y', install_options, :list, wanted].compact)
+      command = [command(:cmd), '-d', '0', '-e', error_level, '-y', install_options, :list, wanted].compact
+      output = execute(command, :failonfail => false, :combine => false)
+
+      if output.exitstatus != 0
```

Comment:
I can parse the output of `yum repolist enabled -v` and only run `yum makecache` when a repo is expired (comparing now to last plus metadata_expire seconds).
```
[root@localhost ~]# yum repolist enabled -v | grep 'Repo-name\|expire'
Repo-name    : CentOS-7 - Base
Repo-expire  : 21,600 second(s) (last: Wed Mar  8 19:01:59 2017)
Repo-name    : CentOS-7 - Extras
Repo-expire  : 21,600 second(s) (last: Wed Mar  8 19:02:00 2017)
Repo-name    : CentOS-7 - Updates
Repo-expire  : 21,600 second(s) (last: Wed Mar  8 19:02:01 2017)
```
I think we can even use `yum makecache fast --disablerepo='*' --enablerepo=<repo>` to target the specific repos which are expired.

Parsing output may have portability concerns. Pulling this programmatically via the yum python library was pretty easy.
```
[root@localhost ~]# python
Python 2.7.5 (default, Aug 18 2016, 15:58:25) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-4)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import yum
>>> y = yum.YumBase()
>>> for repo in y.repos.listEnabled():
...   print repo
...   print repo.metadataCurrent()
... 
Loaded plugins: fastestmirror
base
True
extras
True
localrepo
False
updates
True
```
Do either of these solutions sound reasonable? If the latter, is there a standard/desired way to include a python helper in a ruby project (I'm not aware of comprehensive yum library for ruby).